universal-application-tool-0.0.1/app/services/program/ProgramService.java

```diff
@@ -1,17 +1,88 @@
+  /**
+   * List all programs.
+   *
+   * @return a list of {@link ProgramDefinition}s
+   */
+  ImmutableList<ProgramDefinition> listProgramDefinitions();
 
-    /** Get the definition for a given program. */
-    public ProgramDefinition getProgramDefinition(String id);
+  /**
+   * Get the definition for a given program.
+   *
+   * @param id the ID of the program to retrieve
+   * @return an {@link Optional} of the {@link ProgramDefinition} for the given ID if it exists, or
+   *     {@code Optional.empty()} if not
+   */
+  Optional<ProgramDefinition> getProgramDefinition(long id);
 
-    /** Create a new program. */
-    public ProgramDefinition createProgram(String name, String description);
+  /**
+   * Create a new program.
+   *
+   * @param name a name for this program
+   * @param description the description of what the program provides
+   * @return the {@link ProgramDefinition} that was created
+   */
+  ProgramDefinition createProgram(String name, String description);
 
+  /**
+   * Adds a {@link BlockDefinition} to the given program.
+   *
+   * @param programId the ID of the program to update
+   * @param blockName a name for the block to add
+   * @param blockDescription a description of what the questions in the block address
+   * @return the updated {@link ProgramDefinition}
+   */
+  ProgramDefinition addBlockToProgram(long programId, String blockName, String blockDescription);
+
+  /**
+   * Insert a new {@link BlockDefinition} at a specific index for a given program.
+   *
+   * @param programId the ID of the program to update
+   * @param blockIndex the index in the block list for the new block
+   * @param blockName a name for the block to add
+   * @param blockDescription a description of what the questions in the block address
+   * @return the updated {@link ProgramDefinition}
+   */
+  ProgramDefinition addBlockToProgram(
+      long programId, int blockIndex, String blockName, String blockDescription);
+
+  /**
+   * Update a {@link BlockDefinition} with a set of questions.
+   *
+   * @param programId the ID of the program to update
+   * @param blockIndex the index of the block to update
+   * @param questionIds an {@link ImmutableList} of questions for the block
+   * @return the updated {@link ProgramDefinition}
+   */
+  ProgramDefinition setBlockQuestions(
+      long programId, int blockIndex, ImmutableList<String> questionIds);
+
+  /**
+   * Set the hide {@link Predicate} for a block. This predicate describes under what conditions the
+   * block should be hidden from an applicant filling out the program form.
+   *
+   * @param programId the ID of the program to update
+   * @param blockIndex the index of the block to update
+   * @param predicate the {@link Predicate} for hiding the block
+   * @return the updated {@link ProgramDefinition}
+   */
+  ProgramDefinition setBlockHidePredicate(long programId, int blockIndex, Predicate predicate);
```

Comment:
Thought (blocking): possible to reference the block by ID rather than index?

Reply:
I don't think the BlockDefinition has an ID yet - is that something we want?

Reply:
I guess a larger discussion is "are BlockDefinitions their own model / db object, or are they included in ProgramDefinition"?

Reply:
I think they're child resources of the program. Giving them their own ID is convenient for updates and generating URLs imo. It *could* be done using the index but I don't think that's a good practice.

Reply:
Makes sense to me - I'll make sure we add the ID field on BlockDefinition